consensus/core/src/dag_state.rs

```diff
@@ -202,6 +202,48 @@ impl DagState {
             .collect()
     }
 
+    pub(crate) fn contains_block_in_cache_or_store(
+        &self,
+        block_ref: &BlockRef,
+    ) -> ConsensusResult<bool> {
+        let blocks = self.contains_blocks_in_cache_or_store(vec![*block_ref])?;
+        Ok(blocks.first().cloned().expect("Result should be present"))
+    }
+
+    /// Checks whether the required blocks are in cache, if exist, or otherwise will check in store. The method is not caching
+    /// back the results, so its expensive if keep asking for cache missing blocks.
+    pub(crate) fn contains_blocks_in_cache_or_store(
+        &self,
+        block_refs: Vec<BlockRef>,
+    ) -> ConsensusResult<Vec<bool>> {
+        let mut blocks = Vec::with_capacity(block_refs.len());
+        let mut missing = Vec::new();
+        for (index, block_ref) in block_refs.into_iter().enumerate() {
+            if self.cached_refs[block_ref.author].contains(&block_ref) {
```

Comment:
Very important assumption is that any requested `BlockRef` here is already validated against the committee otherwise this code will blow up. That should be done as part of the block verification process.